api/client/tunneldialer.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/proxy"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+)
+
+// TunnelAuthDialer connects to the Auth Server through the reverse tunnel.
+type TunnelAuthDialer struct {
+	// ProxyAddr is the address of the proxy
+	ProxyAddr string
+	// ClientConfig is SSH tunnel client config
+	ClientConfig *ssh.ClientConfig
+}
+
+// DialContext dials auth server via SSH tunnel
+func (t *TunnelAuthDialer) DialContext(ctx context.Context, network string, _ string) (net.Conn, error) {
```

Comment:
is the address passed as an argument here different from `t.ProxyAddr`?

Reply:
For the api client's use case, the addr and t.ProxyAddr are the same, but it appears that that's not true for other instances of `TunnelAuthDialer`. I'll update `DialContext` to work with both.